### PR TITLE
overlord: add kernel rollback accross reboots manager test and fixes

### PR DESCRIPTION
**Pull request by mvo5**
This adds a test for kernel rollbacks accross reboots into the
managers test suite and fixes the behaviour. It seems like the
old code was not triggering a proper error and we did not notice
this in spread because we cannot replace an asserted kernel with
an unasserted so this test was incomplete.

Based on https://github.com/snapcore/snapd/pull/7649

### CODE REVIEW

File: bootloader/bootloadertest/bootloadertest.go [hunk 1]
```diff
@@ -97,3 +97,13 @@ func (b *MockBootloader) SetBootKernel(kernel string) {
 func (b *MockBootloader) SetBootBase(base string) {
 	b.SetBootVars(map[string]string{"snap_core": base})
 }
+
+func (b *MockBootloader) SetTrying() {
+	b.BootVars["snap_mode"] = "trying"
```
Comment (pedronis) [inline bootloader/bootloadertest/bootloadertest.go]: should we check that snap_mode is try ?

File: bootloader/bootloadertest/bootloadertest.go [hunk 2]
```diff
@@ -97,3 +97,13 @@ func (b *MockBootloader) SetBootKernel(kernel string) {
 func (b *MockBootloader) SetBootBase(base string) {
 	b.SetBootVars(map[string]string{"snap_core": base})
 }
+
+func (b *MockBootloader) SetTrying() {
```
Comment (pedronis) [inline bootloader/bootloadertest/bootloadertest.go]: this should probably have Reboot in the name (also for symmetry with the next one),  SetTryingDuringReboot ?

File: bootloader/bootloadertest/bootloadertest.go [hunk 3]
```diff
@@ -97,3 +97,13 @@ func (b *MockBootloader) SetBootKernel(kernel string) {
...
+	b.BootVars["snap_mode"] = "trying"
+}
+
+func (b *MockBootloader) SetRollbackAccrossReboot() {
+	b.BootVars["snap_mode"] = ""
```
[the remark this answers is not in the document]
Comment (pedronis) [reply bootloader/bootloadertest/bootloadertest.go]: also should we check that snap_mode was try again ?

File: overlord/devicestate/devicemgr.go [hunk 4]
```diff
@@ -429,6 +429,12 @@ func (m *DeviceManager) ensureSeedYaml() error {
 	return nil
 }
 
+// ResetBootOk is only useful for integration testing
+func (m *DeviceManager) ResetBootOk() {
```
Comment (pedronis) [inline overlord/devicestate/devicemgr.go]: thank you

File: overlord/managers_test.go [hunk 5]
```diff
@@ -3498,6 +3523,27 @@ version: 2.0`
 	c.Assert(tasks, HasLen, i+1)
 }
 
+func (ms *mgrsSuite) mockSuccessfulSystemRestart(c *C, bloader *bootloadertest.MockBootloader) {
```
Comment (pedronis) [inline overlord/managers_test.go]: should we use /SystemRestart/Reboot/ ?

Comment (pedronis) [reply overlord/managers_test.go]: they seem important enough helper to go before the first tests using them, not after, but just my 2c

Comment (pedronis) [reply overlord/managers_test.go]: we can also use st.Restarting() to check whether there is a restart of the right kind going on

File: overlord/managers_test.go [hunk 6]
```diff
@@ -1598,6 +1598,33 @@ type: os
 
 }
 
+func (s *mgrsSuite) mockSuccessfulReboot(c *C, bloader *bootloadertest.MockBootloader) {
+	st := s.o.State()
+	restarting, _ := st.Restarting()
```
Comment (pedronis) [inline overlord/managers_test.go]: we can also look at the 2nd return value and check that is a RestartSystem ?

File: overlord/managers_test.go [hunk 7]
```diff
@@ -1598,6 +1598,33 @@ type: os
...
+	s.o.DeviceManager().ResetBootOk()
+	st.Unlock()
+	defer st.Lock()
+	err = s.o.DeviceManager().Ensure()
+	c.Assert(err, IsNil)
+}
+
+func (s *mgrsSuite) mockRollbackAccrossReboot(c *C, bloader *bootloadertest.MockBootloader) {
+	st := s.o.State()
+	restarting, _ := st.Restarting()
```
Comment (pedronis) [inline overlord/managers_test.go]: same

File: overlord/managers_test.go [hunk 8]
```diff
@@ -1598,6 +1598,33 @@ type: os
...
+	state.MockRestarting(st, state.RestartUnset)
+	err := bloader.SetRollbackAcrossReboot()
+	c.Assert(err, IsNil)
+	s.o.DeviceManager().ResetBootOk()
+	st.Unlock()
+	s.o.Settle(settleTimeout)
+	st.Lock()
+}
+
 func (s *mgrsSuite) TestInstallKernelSnapUpdatesBootloader(c *C) {
```
Comment (pedronis) [inline overlord/managers_test.go]: now that we have gadget assets updates this test name is slightly misleading, maybe  UpdatesBootloaderEnv ?  and we should look if there are more test with such names...

File: overlord/snapstate/snapstate.go [hunk 9]
```diff
@@ -434,25 +430,42 @@ func WaitRestart(task *state.Task, snapsup *SnapSetup) (err error) {
...
+		// case we need to catch this and error accordingly
+		current, err := boot.GetCurrentBoot(snapsup.Type)
 		if err == boot.ErrBootNameAndRevisionNotReady {
 			return &state.Retry{After: 5 * time.Second}
 		}
 		if err != nil {
 			return err
 		}
-
+		snapInfo, err := snap.ReadInfo(snapsup.InstanceName(), snapsup.SideInfo)
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go]: do you remember why we read the full snap? the Revision we use below is inside snapup.SideInfo anyway ?

Comment (mvo5) [reply overlord/snapstate/snapstate.go]: I think there is no good reason and it makes the mocks more complicated so I will simplify this.

**Pull request merged**